### ISSUE TITLE
rancher-kubernetes-agent flapping after upgrade from 1.5.7 to 1.6.6-rc3 w/github auth

### ISSUE THREAD

**Issue by moelsayed** [rancher/rancher#9481]
**Rancher versions:**
rancher/server: 1.6.6-rc3

**Infrastructure Stack versions:**
kubernetes (if applicable): rancher/k8s:v1.7.2-rancher1

**Docker version: (`docker version`,`docker info` preferred)**
1.12.6
**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
RHEL 7.3
**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
AWS
**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
single node, internal DB
**Environment Template: (Cattle/Kubernetes/Swarm/Mesos)**
k8s
**Steps to Reproduce:**
* install rancher server 1.5.7
* enable github auth
* deploy k8s
* upgrade to 1.6.6-rc3
* upgrade network-services
* upgrade ipsec
* upgrade healthcheck
* upgrade k8s

**Results:**
rancher-kubernetes-agent keeps flapping with the following error log:
```
time="2017-07-26T01:57:12Z" level=error msg="Couldn't retrieve list of objects, status code: 401" 
time="2017-07-26T01:57:12Z" level=error msg="kubernetes Sync and stream listener exited with error: Couldn't retrieve list of objects, status code: 401" 
time="2017-07-26T01:57:12Z" level=info msg=Exiting. 
```
k8s api log:
```
7/26/2017 4:10:10 AME0726 02:10:10.091487       1 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, [invalid bearer token, Post http://rancher-kubernetes-auth/: dial tcp 10.42.142.57:80: i/o timeout]]]
7/26/2017 4:10:10 AME0726 02:10:10.095157       1 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, [invalid bearer token, Post http://rancher-kubernetes-auth/: dial tcp 10.42.142.57:80: i/o timeout]]]
7/26/2017 4:10:10 AME0726 02:10:10.100311       1 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, [invalid bearer token, Post http://rancher-kubernetes-auth/: dial tcp 10.42.142.57:80: i/o timeout]]]
7/26/2017 4:10:10 AME0726 02:10:10.110473       1 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, [invalid bearer token, Post http://rancher-kubernetes-auth/: dial tcp 10.42.142.57:80: i/o timeout]]]
7/26/2017 4:10:10 AME0726 02:10:10.111532       1 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, [invalid bearer token, Post http://rancher-kubernetes-auth/: dial tcp 10.42.142.57:80: i/o timeout]]]
7/26/2017 4:10:12 AME0726 02:10:12.668870       1 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, [invalid bearer token, Post http://rancher-kubernetes-auth/: dial tcp 10.42.142.57:80: i/o timeout]]]
7/26/2017 4:10:15 AME0726 02:10:15.558076       1 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, [invalid bearer token, Post http://rancher-kubernetes-auth/: dial tcp 10.42.142.57:80: i/o timeout]]]
7/26/2017 4:10:15 AME0726 02:10:15.733214       1 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, [invalid bearer token, Post http://rancher-kubernetes-auth/: dial tcp 10.42.142.57:80: i/o timeout]]]
7/26/2017 4:10:15 AME0726 02:10:15.734728       1 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, [invalid bearer token, Post http://rancher-kubernetes-auth/: dial tcp 10.42.142.57:80: i/o timeout]]]
7/26/2017 4:10:15 AME0726 02:10:15.864643       1 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, [invalid bearer token, Post http://rancher-kubernetes-auth/: dial tcp 10.42.142.57:80: i/o timeout]]]
7/26/2017 4:10:16 AME0726 02:10:16.681205       1 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, [invalid bearer token, Post http://rancher-kubernetes-auth/: dial tcp 10.42.142.57:80: i/o timeout]]]
```

**Comment by moelsayed** [comment]
After restarting the rancher-kubernetes-auth service. k8s completed the upgrade.

However, kubectl-shell container didn't work. It wasn't able to resolve DNS.

**Comment by leodotcloud** [comment]
**Dangling veth issue:**

virtual ethernet interfaces (veth) interfaces are created and destroyed when containers start and stop on a host. Sometimes, these veths are not cleaned up correctly after a container is stopped or destroyed. These dangling veths cause a variety of networking problems.

In our experience, we have seen this issue on RHEL/CentOS hosts more often than other operating systems like Ubuntu.

Also see: https://github.com/moby/moby/issues/31937

**How to detect dangling veths:**

Check if are getting ARP replies from two different sources for the same IP address.

`apt-get update && apt-get install arping`

Example:

```
root@bef6cd4f7380:/# arping 10.42.217.109
ARPING 10.42.217.109
42 bytes from 02:77:58:06:5d:c9 (10.42.217.109): index=0 time=13.823 msec
60 bytes from 02:77:58:df:71:f6 (10.42.217.109): index=1 time=13.849 msec
42 bytes from 02:77:58:06:5d:c9 (10.42.217.109): index=2 time=10.936 msec
60 bytes from 02:77:58:df:71:f6 (10.42.217.109): index=3 time=10.960 msec
42 bytes from 02:77:58:06:5d:c9 (10.42.217.109): index=4 time=8.928 msec
60 bytes from 02:77:58:df:71:f6 (10.42.217.109): index=5 time=8.952 msec
```

**How we fixed this issue:**

Reference: https://github.com/rancher/rancher/issues/7955
Fixed in `rancher/server:v1.6.3` [Specifically in Network Services v0.2.2, network-manager:v0.7.3]

In network-manager versions < v0.7.3, the veth name on the host was generated using random characters which made it difficult to associate the relationship with the running container. As part of the fix, we changed the naming convention to use `vethr` followed by first few characters of the docker container ID. A new goroutine has been added in network-manager which monitors the veths in the host and deletes the dangling ones if found.

But in an upgraded setup, you could have containers started before the new network-manager logic kicks in. So we cannot figure out the dangling veths from the past, but we can track them going forward. (We actually tried to address this upgrade scenario as well, but we had to revert those changes as we couldn't get this to work reliably on all cloud providers. https://github.com/rancher/plugin-manager/commit/fb32a9d234f28b7c2a6b6156c47ce6858f574587)


**Workaround for upgraded setups:**
Unfortunately there is no safe and reliable way to figure out the dangling veths in an upgraded setup. So the workaround to fix the hosts which have dangling veths is to reboot the host. Since we cannot figure out which hosts have the dangling veth, a reboot of all the hosts is advised.
